clippy_lints/src/needless_continue.rs

```diff
@@ -361,24 +401,79 @@ fn suggestion_snippet_for_continue_inside_else(cx: &EarlyContext<'_>, data: &Lin
     )
 }
 
-fn check_and_warn(cx: &EarlyContext<'_>, expr: &ast::Expr) {
-    if_chain! {
-        if let ast::ExprKind::Loop(loop_block, ..) = &expr.kind;
-        if let Some(last_stmt) = loop_block.stmts.last();
-        if let ast::StmtKind::Expr(inner_expr) | ast::StmtKind::Semi(inner_expr) = &last_stmt.kind;
-        if let ast::ExprKind::Continue(_) = inner_expr.kind;
-        then {
-            span_lint_and_help(
-                cx,
-                NEEDLESS_CONTINUE,
-                last_stmt.span,
-                MSG_REDUNDANT_CONTINUE_EXPRESSION,
-                None,
-                DROP_CONTINUE_EXPRESSION_MSG,
-            );
+fn check_last_stmt_in_if<F>(b: &ast::Expr, func: &F)
+where
+    F: Fn(Option<&ast::Label>, Span),
+{
+    match &b.kind {
+        ast::ExprKind::If(_, then_block, else_block) => {
+            check_last_stmt(then_block, func);
+            if let Some(else_block) = else_block {
+                check_last_stmt_in_if(else_block, func);
+            }
+        },
+        ast::ExprKind::Continue(..) => {
+            unreachable!()
+        },
+        ast::ExprKind::Block(b, _) => {
+            check_last_stmt(b, func);
+        },
+        _ => {},
+    }
+}
+
+fn check_last_stmt<F>(b: &Block, func: &F)
+where
+    F: Fn(Option<&ast::Label>, Span),
+{
+    if let Some(last_stmt) = b.stmts.last() &&
+       let ast::StmtKind::Expr(inner_expr) | ast::StmtKind::Semi(inner_expr) = &last_stmt.kind {
+        match &inner_expr.kind {
+            ast::ExprKind::Continue(continue_label) => {
+                func(continue_label.as_ref(), last_stmt.span);
+            },
+            ast::ExprKind::If(_, _, _) => {
+                check_last_stmt_in_if(inner_expr, func);
+            }
+            ast::ExprKind::Match(_, arms) => {
+                for arm in arms {
+                    match &arm.body.kind {
+                        ast::ExprKind::Continue(continue_label) => {
+                            func(continue_label.as_ref(), arm.body.span);
+                        }
+                        ast::ExprKind::Block(b, _) => {
+                            check_last_stmt(b, func);
+
+                        }
+                        _ => {}
+                    }
+
+                }
+            }
+            ast::ExprKind::Block(b, _) => {
+                check_last_stmt(b, func);
+            }
```

Comment:
It would be good to flatten this since it's currently handling the same expressions in a few different places, you could split it into two functions `check_last_stmt` and `check_last_stmt_in_expr` (replacing `check_last_stmt_in_if`)

e.g. roughly

```rust
fn check_last_stmt_in_expr(..) {
    match &expr.kind {
        ExprKind::Continue(continue_label) => {
            // lint, could inline it since now it's only happening in one place
        },
        ExprKind::If(_, then_block, else_block) => {
            check_last_stmt(..);
            if let Some(else_block) = else_block {
                check_last_stmt_in_expr(..);
            }
        },
        ExprKind::Match(_, arms) => {
            for arm in arms {
                check_last_stmt_in_expr(..);
            }
        },
        ExprKind::Block(b, _) => {
            check_last_stmt(..);
        },
        _ => {},
    }
}

fn check_last_stmt(..) {
    if let Some(last_stmt) = b.stmts.last()
        && let ast::StmtKind::Expr(expr) | ast::StmtKind::Semi(expr) = &last_stmt.kind
    {
        check_last_stmt_in_expr(expr, ..);
    }
}
```